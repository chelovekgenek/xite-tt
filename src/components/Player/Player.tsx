import React, { useRef, useCallback, useEffect } from "react"
import { connect } from "react-redux"
import { Player as VideoPlayer, ControlBar } from "video-react"

import { TAppState } from "store/entities"
import { getExpandedSelected, VideoSelectedActions } from "store/entities/video"

interface IStateProps {
  selected: ReturnType<typeof getExpandedSelected>
}
interface IDispatchProps {
  next: typeof VideoSelectedActions.setNext
}
interface IProps extends IStateProps, IDispatchProps {}

export const Player: React.FC<IProps> = ({ selected, next }) => {
  const player = useRef(null)
  const handleSubscription = useCallback(
    (nextState, prevState) => {
      if (nextState.ended && !prevState.ended) {
        next()
      }
    },
    [next],
  )
  useEffect(() => {
    if (player.current) {
      ;(player.current as any).subscribeToStateChange(handleSubscription) // for real life app it has to be typed
    }
  })
  if (!selected) {
    return null
  }
  return (
    <React.Fragment>
      <h4>
        {selected.title} | {selected.artist}
      </h4>
      <VideoPlayer ref={player} autoPlay key={selected.id} fluid={false} width={"100%"} height={"calc(100% - 2rem)"}>
        <source src={selected.src} />
        <ControlBar autoHide={false} />
      </VideoPlayer>
    </React.Fragment>
  )
}

export default connect<IStateProps, IDispatchProps, {}, TAppState>(
  state => ({ selected: getExpandedSelected(state) }),
  {
    next: VideoSelectedActions.setNext,
  },
)(Player)
