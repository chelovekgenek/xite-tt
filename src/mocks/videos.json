[
  {
    "id": "1",
    "title": "asd",
    "artist": "dsa",
    "src": "https://media.w3.org/2010/05/sintel/trailer_hd.mp4"
  }
]
