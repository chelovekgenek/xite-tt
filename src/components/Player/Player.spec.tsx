import React from "react"
import { shallow } from "enzyme"

import videos from "mocks/videos.json"
import { Player } from "./Player"

describe("Player", () => {
  it("renders correctly", () => {
    const comp = shallow(<Player selected={videos[0]} next={jest.fn() as any} />)
    expect(comp).toMatchSnapshot()
  })
  it("shouldn't render if nothing is selected", () => {
    const comp = shallow(<Player selected={undefined} next={jest.fn() as any} />)
    expect(comp).toMatchSnapshot()
  })
})
