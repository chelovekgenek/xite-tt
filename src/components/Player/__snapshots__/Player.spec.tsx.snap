// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player renders correctly 1`] = `
<Fragment>
  <h4>
    asd
     | 
    dsa
  </h4>
  <Player
    aspectRatio="auto"
    autoPlay={true}
    fluid={false}
    height="calc(100% - 2rem)"
    key="1"
    muted={false}
    playsInline={false}
    preload="auto"
    width="100%"
  >
    <source
      src="https://media.w3.org/2010/05/sintel/trailer_hd.mp4"
    />
    <ControlBar
      autoHide={false}
      disableCompletely={false}
    />
  </Player>
</Fragment>
`;

exports[`Player shouldn't render if nothing is selected 1`] = `""`;
